browser/brave_rewards/creator_detection_script_injector.cc

```diff
@@ -0,0 +1,186 @@
+     {"m.youtube.com", IDR_CREATOR_DETECTION_YOUTUBE_BUNDLE_JS}});
+
+std::string LoadScriptResource(int id) {
+  auto& bundle = ui::ResourceBundle::GetSharedInstance();
+  if (bundle.IsGzipped(id)) {
+    return bundle.LoadDataResourceString(id);
+  }
+  return std::string(bundle.GetRawDataResource(id));
+}
+
+std::optional<std::string> GetDetectionScript(content::RenderFrameHost* rfh) {
+  // Only run scripts for the main frame.
+  if (!rfh || !rfh->IsInPrimaryMainFrame()) {
+    return std::nullopt;
+  }
+
+  // Only run scripts if the creator detection feature is enabled.
+  if (!base::FeatureList::IsEnabled(
+          features::kPlatformCreatorDetectionFeature)) {
+    return std::nullopt;
+  }
+
+  auto* profile = Profile::FromBrowserContext(rfh->GetBrowserContext());
+
+  // Only run scripts if the Rewards service is available for this profile.
+  if (!IsSupportedForProfile(profile)) {
+    return std::nullopt;
+  }
+
+  // Only run scripts if the user has enabled Brave Rewards.
+  if (!profile || !profile->GetPrefs()->GetBoolean(prefs::kEnabled)) {
+    return std::nullopt;
+  }
+
+  // Only run scripts for known "media platform" sites.
+  GURL url = rfh->GetLastCommittedURL();
+  if (!IsMediaPlatformURL(url)) {
+    return std::nullopt;
+  }
+
+  // Only run scripts when there is an exact hostname match.
+  auto iter = kScriptMap.find(url.host_piece());
+  if (iter == kScriptMap.end()) {
+    return std::nullopt;
+  }
+
+  return LoadScriptResource(iter->second);
+}
+
+}  // namespace
+
+CreatorDetectionScriptInjector::CreatorDetectionScriptInjector() = default;
+CreatorDetectionScriptInjector::~CreatorDetectionScriptInjector() = default;
+
+void CreatorDetectionScriptInjector::MaybeInjectScript(
+    content::RenderFrameHost* rfh) {
+  injector_.reset();
+  injector_host_ = rfh;
+  last_url_ = GURL();
+
+  auto script_source = GetDetectionScript(injector_host_);
+  if (!script_source) {
+    return;
+  }
+
+  injector_host_->GetRemoteAssociatedInterfaces()->GetInterface(&injector_);
+
+  bool verbose_logging =
+      base::FeatureList::IsEnabled(features::kVerboseLoggingFeature);
+
+  // Add a `braveRewards` global variable to the isolated world.
+  std::string initializer =
+      base::StrCat({"self.braveRewards = { verboseLogging: ",
+                    verbose_logging ? "true" : "false", " }"});
+  ExecuteScript(initializer, base::DoNothing());
+
+  // Execute the detection script. It must set `braveRewards.detectCreator` to a
+  // function. That function will be called by `DetectCreator`.
+  ExecuteScript(script_source.value(), base::DoNothing());
+}
+
+void CreatorDetectionScriptInjector::DetectCreator(
+    content::RenderFrameHost* rfh,
+    DetectCreatorCallback callback) {
+  if (!rfh || rfh != injector_host_.get() || !injector_.is_bound()) {
+    std::move(callback).Run(std::nullopt);
+    return;
+  }
+
+  // Do not call the detection routine if it has already been run for this URL.
```

Comment:
Seems like this is something that the tab helper should be responsible for?

Reply:
Moved this check to `RewardsTabHelper`.